### ISSUE TITLE
[homematic]: HMIP, ENOCEAN, CUxD, Systemvariables are not discovered - not working

### ISSUE THREAD

**Issue by MHerbst** [openhab/openhab-addons#8651]
After upgrading to OH2.5.9 and Zulu “1.8.0_265” the homematic binding is only discovering homematic devices - HMIP, ENOCEAN, CUxD, and systemvariables are not any more discovered.

Furthermore even if the device is “online” actions are not recognised in openhab anymore.

````
###############################################################################
############### openhab #####################################################
###############################################################################
Ip = 192.168.XXX.XX
Release = Raspbian GNU/Linux 10 (buster)
Kernel = Linux 5.4.51-v7l+
Platform = Raspberry Pi 4 Model B Rev 1.2
Uptime = 2 day(s). 2:45:9
CPU Usage = 1.75% avg over 4 cpu(s) (4 core(s) x 1 socket(s))
CPU Load = 1m: 0.65, 5m: 0.33, 15m: 0.26
Memory = Free: 0.96GB (26%), Used: 2.81GB (74%), Total: 3.78GB
Swap = Free: 2.58GB (100%), Used: 0.00GB (0%), Total: 2.58GB
Root = Free: 6.67GB (48%), Used: 7.03GB (52%), Total: 14.32GB
Updates = 0 apt updates available.
Sessions = 2 session(s)
Processes = 138 running processes of 32768 maximum processes
###############################################################################
````
When I try to discover things with the homematic binding, i get following error message:

````
2020-09-30 10:55:03.775 [WARN ] [very.HomematicDeviceDiscoveryService] - Failed to set Homematic controller in install mode
java.io.IOException: -1 setInstallMode: unknown.method name (sending setInstallMode()
true
60
1
)
at org.openhab.binding.homematic.internal.communicator.parser.RpcResponseParser.parse(RpcResponseParser.java:51) ~[bundleFile:?]
at org.openhab.binding.homematic.internal.communicator.client.BinRpcClient.sendMessage(BinRpcClient.java:82) ~[bundleFile:?]
at org.openhab.binding.homematic.internal.communicator.client.BinRpcClient.sendMessage(BinRpcClient.java:94) ~[bundleFile:?]
at org.openhab.binding.homematic.internal.communicator.client.BinRpcClient.sendMessage(BinRpcClient.java:94) ~[bundleFile:?]
at org.openhab.binding.homematic.internal.communicator.client.BinRpcClient.sendMessage(BinRpcClient.java:94) ~[bundleFile:?]
at org.openhab.binding.homematic.internal.communicator.client.BinRpcClient.sendMessage(BinRpcClient.java:70) ~[bundleFile:?]
at org.openhab.binding.homematic.internal.communicator.client.RpcClient.setInstallMode(RpcClient.java:394) ~[bundleFile:?]
at org.openhab.binding.homematic.internal.communicator.AbstractHomematicGateway.setInstallMode(AbstractHomematicGateway.java:597) ~[bundleFile:?]
at org.openhab.binding.homematic.internal.discovery.HomematicDeviceDiscoveryService.enableInstallMode(HomematicDeviceDiscoveryService.java:109) [bundleFile:?]
at org.openhab.binding.homematic.internal.discovery.HomematicDeviceDiscoveryService.startScan(HomematicDeviceDiscoveryService.java:90) [bundleFile:?]
at org.eclipse.smarthome.config.discovery.AbstractDiscoveryService.startScan(AbstractDiscoveryService.java:209) [bundleFile:?]
at org.eclipse.smarthome.config.discovery.internal.DiscoveryServiceRegistryImpl.startScan(DiscoveryServiceRegistryImpl.java:382) [bundleFile:?]
at org.eclipse.smarthome.config.discovery.internal.DiscoveryServiceRegistryImpl.startScans(DiscoveryServiceRegistryImpl.java:358) [bundleFile:?]
at org.eclipse.smarthome.config.discovery.internal.DiscoveryServiceRegistryImpl.startScan(DiscoveryServiceRegistryImpl.java:216) [bundleFile:?]
at org.eclipse.smarthome.io.rest.core.internal.discovery.DiscoveryResource.scan(DiscoveryResource.java:97) [bundleFile:?]
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_265]
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_265]
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_265]
at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_265]
at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java:81) [bundleFile:?]
at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:144) [bundleFile:?]
at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:161) [bundleFile:?]
at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:160) [bundleFile:?]
at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:99) [bundleFile:?]
at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:389) [bundleFile:?]
at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:347) [bundleFile:?]
at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:102) [bundleFile:?]
at org.glassfish.jersey.server.ServerRuntime$2.run(ServerRuntime.java:326) [bundleFile:?]
at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271) [bundleFile:?]
at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267) [bundleFile:?]
at org.glassfish.jersey.internal.Errors.process(Errors.java:315) [bundleFile:?]
at org.glassfish.jersey.internal.Errors.process(Errors.java:297) [bundleFile:?]
at org.glassfish.jersey.internal.Errors.process(Errors.java:267) [bundleFile:?]
at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:317) [bundleFile:?]
at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:305) [bundleFile:?]
at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1154) [bundleFile:?]
at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:473) [bundleFile:?]
at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:427) [bundleFile:?]
at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:388) [bundleFile:?]
at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:341) [bundleFile:?]
at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:228) [bundleFile:?]
at com.eclipsesource.jaxrs.publisher.internal.ServletContainerBridge.service(ServletContainerBridge.java:76) [bundleFile:?]
at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:852) [bundleFile:9.4.20.v20190813]
at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:544) [bundleFile:9.4.20.v20190813]
at org.ops4j.pax.web.service.jetty.internal.HttpServiceServletHandler.doHandle(HttpServiceServletHandler.java:71) [bundleFile:?]
at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143) [bundleFile:9.4.20.v20190813]
at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:536) [bundleFile:9.4.20.v20190813]
at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127) [bundleFile:9.4.20.v20190813]
at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235) [bundleFile:9.4.20.v20190813]
at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1581) [bundleFile:9.4.20.v20190813]
at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233) [bundleFile:9.4.20.v20190813]
at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1307) [bundleFile:9.4.20.v20190813]
at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext.doHandle(HttpServiceContext.java:293) [bundleFile:?]
at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188) [bundleFile:9.4.20.v20190813]
at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:482) [bundleFile:9.4.20.v20190813]
at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1549) [bundleFile:9.4.20.v20190813]
at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186) [bundleFile:9.4.20.v20190813]
at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1204) [bundleFile:9.4.20.v20190813]
at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) [bundleFile:9.4.20.v20190813]
at org.ops4j.pax.web.service.jetty.internal.JettyServerHandlerCollection.handle(JettyServerHandlerCollection.java:80) [bundleFile:?]
at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127) [bundleFile:9.4.20.v20190813]
at org.eclipse.jetty.server.Server.handle(Server.java:494) [bundleFile:9.4.20.v20190813]
at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:374) [bundleFile:9.4.20.v20190813]
at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:268) [bundleFile:9.4.20.v20190813]
at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311) [bundleFile:9.4.20.v20190813]
at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103) [bundleFile:9.4.20.v20190813]
at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117) [bundleFile:9.4.20.v20190813]
at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336) [bundleFile:9.4.20.v20190813]
at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313) [bundleFile:9.4.20.v20190813]
at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171) [bundleFile:9.4.20.v20190813]
at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129) [bundleFile:9.4.20.v20190813]
at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:367) [bundleFile:9.4.20.v20190813]
at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:782) [bundleFile:9.4.20.v20190813]
at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:918) [bundleFile:9.4.20.v20190813]
at java.lang.Thread.run(Thread.java:748) [?:1.8.0_265]
````

and

````
2020-09-30 10:55:08.356 [ERROR] [very.HomematicDeviceDiscoveryService] - null
java.lang.NullPointerException: null
at org.openhab.binding.homematic.internal.communicator.virtual.AbstractVirtualDatapointHandler.addDatapoint(AbstractVirtualDatapointHandler.java:75) ~[?:?]
at org.openhab.binding.homematic.internal.communicator.virtual.AbstractVirtualDatapointHandler.addDatapoint(AbstractVirtualDatapointHandler.java:67) ~[?:?]
at org.openhab.binding.homematic.internal.communicator.virtual.FirmwareVirtualDatapointHandler.initialize(FirmwareVirtualDatapointHandler.java:34) ~[?:?]
at org.openhab.binding.homematic.internal.communicator.AbstractHomematicGateway.prepareDevice(AbstractHomematicGateway.java:823) ~[?:?]
at org.openhab.binding.homematic.internal.communicator.AbstractHomematicGateway.loadAllDeviceMetadata(AbstractHomematicGateway.java:414) ~[?:?]
at org.openhab.binding.homematic.internal.discovery.HomematicDeviceDiscoveryService.lambda$0(HomematicDeviceDiscoveryService.java:209) ~[?:?]
at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_265]
at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_265]
at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:1.8.0_265]
at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:1.8.0_265]
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_265]
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_265]
at java.lang.Thread.run(Thread.java:748) [?:1.8.0_265]
````

Before the upgrade the binding (OH2.5.8) worked without any problems.
Thanks for any help

See also: https://community.openhab.org/t/homematic-hmip-enocean-cuxd-systemvariables-are-not-discovered-not-working/105769

**Comment by MHerbst** [comment]
After received a more detailed TRACE log it seems that the problem is caused by a special device:

````
2020-10-03 21:22:01.919 [TRACE] [ommunicator.AbstractHomematicGateway] - Loading metadata for device 'PEQ1418539' of type 'HMW-Generic'
2020-10-03 21:22:01.919 [TRACE] [ommunicator.AbstractHomematicGateway] -   Loading channel HmChannel[number=-1,type=HMW-Generic,initialized=false]
2020-10-03 21:22:01.920 [TRACE] [ommunicator.AbstractHomematicGateway] -     Loading datapoints into channel HmChannel[number=-1,type=HMW-Generic,initialized=false]
2020-10-03 21:22:01.920 [TRACE] [nal.communicator.client.XmlRpcClient] - Client XmlRpcRequest (port 2000):
<?xml version="1.0" encoding="ISO-8859-1"?>
<methodCall><methodName>getParamsetDescription</methodName>
<params><param><value>XXXXXXXX</value></param><param><value>MASTER</value></param></params></methodCall>
2020-10-03 21:22:01.924 [TRACE] [nal.communicator.client.XmlRpcClient] - Client XmlRpcResponse (port 2000):
<?xml version="1.0" encoding="iso-8859-1"?>
<methodResponse><params><param>
	<value><struct></struct></value>
</param></params></methodResponse>

2020-10-03 21:22:01.926 [ERROR] [very.HomematicDeviceDiscoveryService] - null
java.lang.NullPointerException: null
`````

HWM-GENERIC is a special device type that has not data points. This causes a NPE and the NPE then breaks the discovery and further devices are not detected. Also GATEWAY-EXTRAS for the system variables is not created.

**Comment by TheNightmanX** [comment]
I have the same problem and need a fix for OpenHAB 2.5.x

My config:

OpenHAB 2.5.11 release build
Rpi 3, Debian Linux, apt installation
CCU2 firmware 2.55.5 (is current)
CUxD version 2.5 (is current)

I am not experienced in debugging OpenHAB or LINUX (I am only a user and somewhat experienced in OpenHAB configuration). I f I can contribute to a solution, please tell me exactly what I should do and I will help.

**Comment by lsiepel** [comment]
Is this is still an issue with 3.4.0 stable? 
There have been quite some homematic PR's https://github.com/openhab/openhab-addons/pulls?q=is%3Apr+in%3Atitle+homematic+created%3A%3E2020-10-03++is%3Amerged
And some related to NPE's
